crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -966,37 +979,41 @@ impl<'db> TypeInferenceBuilder<'db> {
     ///   - `from .foo import bar` => `tail == "foo"`
     ///   - `from ..foo.bar import baz` => `tail == "foo.bar"`
     fn relative_module_name(
-        &self,
+        &mut self,
         tail: Option<&str>,
         level: NonZeroU32,
-    ) -> Result<ModuleName, ModuleResolutionError> {
+        import_from: &ast::StmtImportFrom,
+    ) -> Option<ModuleName> {
```

Comment:
A couple of things feel like they become more awkward about this function as a result of getting rid of the `ModuleResolutionError` enum, FWIW:
- It now has to take `&mut self` rather than `&self`, since we emit diagnostics directly from this method
- It has to be passed the `import_from` node as an additional argument so that it can emit diagnostics directly. This is annoying because the `import_from` node has its own `.level` field, but we don't want that to be used from this method because we only call this method if we've already determined that `level > 0`, which is enforced by the fact that the type signature takes a `NonZeroU32`.
- Because the diagnostic could be emitted in several places from this function, I had to factor it out into a separate helper function

Reply:
> It seems like one way to make things cleaner would be to make the signature just `fn import_from_module_name(&mut self, import_from: &ast::StmtImportFrom) -> Option<ModuleName>` and always call it (i.e. incorporate the outer `match` from the callsite into it.) This method only has one call site so it's pretty much arbitrary where we draw the boundary around it.

I just tried that locally. It _works_... but it still feels messy, however, in my opinion, because it's hard to keep track of where we're emitting tracing logs to note invalid syntax. Since the various errors are handled inside `relative_module_name()`, it makese sense to do the logging inside that method. But the caller of the method has to handle the fact that the object returned from `relative_module_name()` could be `None`, and it feels very strange to discard the `None` and replace it with `Type::Unknown` at the callsite without doing any logging _there_. So you end up having to add a verbose comment to the callsite explaining that you've already done the logging inside the helper function -- at which point, you really start to wonder if this is the best design, or if it wouldn't be better for the helper function to be stateless and propagate errors to the caller for the caller to log tracing messages about 😄

> If you strongly prefer using an error enum, I'm open to sticking with that; I just find it generally simpler and clearer to emit diagnostics where they occur and avoid temporary extra representations.

In general I absolutely agree... but here, it feels like there's just too many branches